internal/definitions/sender_tokenpool.go

```diff
@@ -24,34 +24,91 @@ import (
 	"github.com/hyperledger/firefly-common/pkg/log"
 	"github.com/hyperledger/firefly/internal/coremsgs"
 	"github.com/hyperledger/firefly/pkg/core"
+	"github.com/hyperledger/firefly/pkg/database"
 )
 
-func (bm *definitionSender) DefineTokenPool(ctx context.Context, pool *core.TokenPoolAnnouncement, waitConfirm bool) error {
+func (ds *definitionSender) PublishTokenPool(ctx context.Context, poolNameOrID, networkName string, waitConfirm bool) (pool *core.TokenPool, err error) {
+	if !ds.multiparty {
+		return nil, i18n.NewError(ctx, coremsgs.MsgActionNotSupported)
+	}
 
-	if bm.multiparty {
-		// Map token connector name -> broadcast name
-		if broadcastName, exists := bm.tokenBroadcastNames[pool.Pool.Connector]; exists {
-			pool.Pool.Connector = broadcastName
-		} else {
-			log.L(ctx).Infof("Could not find broadcast name for token connector: %s", pool.Pool.Connector)
-			return i18n.NewError(ctx, coremsgs.MsgInvalidConnectorName, broadcastName, "token")
+	var sender *sendWrapper
+	err = ds.database.RunAsGroup(ctx, func(ctx context.Context) error {
+		if pool, err = ds.assets.GetTokenPoolByNameOrID(ctx, poolNameOrID); err != nil {
+			return err
+		}
+		if networkName != "" {
+			pool.NetworkName = networkName
 		}
 
-		if err := pool.Pool.Validate(ctx); err != nil {
-			return err
+		sender = ds.getTokenPoolSender(ctx, pool)
+		if sender.err != nil {
+			return sender.err
 		}
+		if !waitConfirm {
+			if err = sender.sender.Prepare(ctx); err != nil {
+				return err
+			}
+			if err = ds.database.UpsertTokenPool(ctx, pool, database.UpsertOptimizationExisting); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = sender.send(ctx, waitConfirm)
+	return pool, err
+}
 
-		pool.Pool.Namespace = ""
-		msg, err := bm.sendDefinitionDefault(ctx, pool, core.SystemTagDefinePool, waitConfirm)
-		if msg != nil {
-			pool.Pool.Message = msg.Header.ID
+func (ds *definitionSender) getTokenPoolSender(ctx context.Context, pool *core.TokenPool) *sendWrapper {
+	// Map token connector name -> broadcast name
+	if broadcastName, exists := ds.tokenBroadcastNames[pool.Connector]; exists {
+		pool.Connector = broadcastName
+	} else {
+		log.L(ctx).Infof("Could not find broadcast name for token connector: %s", pool.Connector)
+		return wrapSendError(i18n.NewError(ctx, coremsgs.MsgInvalidConnectorName, broadcastName, "token"))
+	}
+
+	if err := pool.Validate(ctx); err != nil {
```

Comment:
Note to self to go look at Validate, and check if we do a pre-publish check on the uniqueness of the `networkName`, as it seems like we haven't committed to a DB update at this point (at least in the `waitConfirm=true` path).

Reply:
Yea, our pre-publish logic is wildly inconsistent for most definitions. Maybe should be cleaning it up as I go, or should at least come back round on all of them.

Reply:
OK as noted in https://github.com/hyperledger/firefly/pull/1261#discussion_r1195556197 I'll remove the pre-flight upsert from the `waitConfirm=false` path, and add a new pre-flight check on network name uniqueness for all paths.